File: 229-MajorityElementIi/229-MajorityElementIi.py
# Last updated: 07/04/2026, 11:40:04
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        e1 = -float("inf")
        e2 = -float("inf")

        c1 = c2 = 0

        if len(nums) <3 :
            return list(set(nums))
            
    

        
        for i in nums:
            
            if c1 == 0 and i != e2  :
                e1 = i 
                c1 = 1

            elif c2 == 0 and i != e1 :
                e2 = i 
                c2 = 1

            elif i == e1 : c1+=1
            elif i == e2 : c2+=1
            else:
                c1-=1
                c2-=1


        

        m = int(len(nums)/3)+1

        c1 = c2 = 0 

        for i in nums :
            if e1 == i :
                c1 +=1 
            elif e2 ==i :
                c2 +=1
        ans = []
        if c1 >= m :
            ans.append(e1)
        if c2 >= m :
            ans.append(e2) 


        return ans

        
        

File: 229-MajorityElementII/229-MajorityElementII.java
// Last updated: 15/04/2026, 12:56:34
1class Solution {
2    public List<Integer> majorityElement(int[] nums) {
3        final int n = nums.length;
4        int a = 0 , ca = 0;
5        int b = 0 , cb = 0;
6        int m = (n/3)+1;
7        List<Integer> ans= new ArrayList<>();
8        
9        for ( int num :nums){
10            if (num == a){
11                ca++;
12            }
13            else if ( num == b){
14                cb ++;
15            }
16            else if( ca == 0){
17                a = num;
18                ca =1;
19            }
20            else if ( cb ==0){
21                b = num;
22                cb = 1;
23            }
24            else {
25                ca-=1;
26                cb-=1;
27            }
28
29
30        }
31
32
33        ca = 0;
34        cb = 0;
35        for ( int num :nums){
36            if ( num == a ) ca++;
37            else if ( num == b) cb++;
38        }
39        if ( ca >=m) ans.add(a);
40        if ( cb>=m) ans.add(b);
41        return ans;
42
43        
44
45    }
46}